plat/mediatek/mt8173/drivers/crypt/crypt.c

```diff
@@ -0,0 +1,131 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <arch.h>
+#include <arch_helpers.h>
+#include <assert.h>
+#include <debug.h>
+#include <delay_timer.h>
+#include <mmio.h>
+#include <mt8173_def.h>
+
+#define crypt_read32(offset)	\
+	mmio_read_32((uintptr_t)(CRYPT_BASE+((offset) * 4)))
+
+#define crypt_write32(offset, value)    \
+	mmio_write_32((uintptr_t)(CRYPT_BASE + ((offset) * 4)), (uint32_t)value)
+
+#define GET_L32(x) ((uint32_t)(x & 0xffffffff))
+#define GET_H32(x) ((uint32_t)((x >> 32) & 0xffffffff))
+
+#define REG_0 0
+#define REG_4 4
+#define REG_256 256
+#define REG_512 512
+#define REG_513 513
+#define REG_514 514
+#define REG_768 768
+#define REG_769 769
+#define REG_770 770
+#define REG_771 771
+#define REG_772 772
+#define REG_820 820
+
+uint64_t crypt_wait(void)
+{
+	crypt_write32(REG_256, 0);
+	while(crypt_read32(REG_512) == 0);
+	udelay(0x64);
+	crypt_write32(REG_513, crypt_read32(REG_512));
+	crypt_write32(REG_514, 0);
+
+	return 0;
+}
+
+static uint32_t record[4];
+uint64_t crypt_set_hdcp_key_ex(uint64_t x1, uint64_t x2, uint64_t x3)
+{
+	uint32_t i = (uint32_t)x1;
+	uint32_t j = 0;
+
+	if (i < 40) {
+		crypt_write32(REG_4, 0x80ff3800);
+		crypt_write32(REG_0, 0);
+		crypt_write32(REG_0, 0xF);
+		crypt_write32(REG_513, 1);
+		crypt_write32(REG_514, 0);
+
+		crypt_write32(REG_768, 0x70);
+		crypt_write32(REG_769, 0x1C0);
+		crypt_write32(REG_770, 0x30);
+		crypt_write32(REG_771, 0x4);
+		crypt_wait();
+
+		crypt_write32(REG_820 + 4 * i, GET_L32(x2));
+		crypt_write32(REG_820 + 4 * i + 1, GET_H32(x2));
+		crypt_write32(REG_820 + 4 * i + 2, GET_L32(x3));
+		crypt_write32(REG_820 + 4 * i + 3, GET_H32(x3));
+
+		crypt_write32(REG_769, 0);
+
+		crypt_write32(REG_768, 0x20);
+		crypt_write32(REG_771, 0x34 + 4 * i);
+		crypt_write32(REG_772, 0x34 + 4 * i);
+		crypt_wait();
+
+		for (j = 0; j < 4; j++) {
+			crypt_write32(REG_768, 0x71);
+			crypt_write32(REG_769, 0x34 + 4 * i + j);
+			crypt_write32(REG_770, record[j]);
+			crypt_wait();
+		}
+	}
+	record[0] = GET_L32(x2);
+	record[1] = GET_H32(x2);
+	record[2] = GET_L32(x3);
+	record[3] = GET_H32(x3);
+	return 0;
+}
+
+uint64_t crypt_set_hdcp_key_num(uint32_t num)
+{
+	crypt_write32(REG_768, 0x6A);
+	crypt_write32(REG_769, 0x34 + 4 * num);
```

Comment:
There is no input validation of non-secure parameter `num` here. Should this value also be <40 as is the case for `x1` in the previous function?